### PR TITLE
Use hashset to filter validators ids in http_api

### PR DESCRIPTION
**Pull request by 0xalex88**
## Issue Addressed

When requesting validators or validators balances via the http api, the endpoint filters the state validators by iterating over the input array for each state validator.
This causes a noticeable slowdown when passing a few thousand IDs in the query filter (from 2s to 12+ when passing 8k IDs).

## Proposed Changes

Use a `std::collections::HashSet` to know if a validator ID/pubkey has to be included in the response instead of iterating over the array, doing at most 2 HashSet lookups per state validator.

### CODE REVIEW

**Comment by 0xalex88** [comment]
@michaelsproul (tagging you since we've talked about this on discord) I've tested it on a mainnet instance and while there's an increase when filtering 1-100 validators indexes (0.005s to 0.05s) after 100 validators the difference becomes substantial, at 8k indexes it takes ~0.1s vs the original 5s

not sure if relevant or might be the cause of the increase in response time on small number of validators but the original one was 5.1.3-modern and the custom built was just the `docker build` of the Dockerfile without additional flags

**Review by dapplion — APPROVED** [review]
LGTM, no good reason to keep a quadratic operation in the API. This filter change does not alter the ordering of the response 👍

File: beacon_node/http_api/src/validators.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/validators.rs]: Clippy wants to simplify this closure:

```suggestion
                    query_ids.as_ref().map(HashSet::from_iter);
```

**Review by michaelsproul — APPROVED** [review]
LGTM once CI passes

**Comment by michaelsproul** [comment]
@mergifyio queue

**Comment by 0xalex88** [comment]
Sorry seems clippy [failed](https://github.com/sigp/lighthouse/actions/runs/8579465837/job/23591988598#step:5:1401), should I create a new PR?

**Comment by michaelsproul** [comment]
@0xalex88 no it's ok, I amended it and the mergify bot checked that clippy was passing before merging to unstable

**Pull request merged**
